util/dscopy/main.go

```diff
@@ -0,0 +1,253 @@
+	}
+	if len(fromBadgerRepos) == 0 && len(fromMongoUri) == 0 {
+		return fmt.Errorf("source not specified")
+	}
+	if len(toBadgerRepos) != 0 && len(toMongoUri) != 0 {
+		return fmt.Errorf("multiple destinations specified")
+	}
+	if len(toBadgerRepos) == 0 && len(toMongoUri) == 0 {
+		return fmt.Errorf("destination not specified")
+	}
+
+	var from, to ds.Datastore
+	var err error
+	if len(fromBadgerRepos) != 0 {
+		path := filepath.Join(fromBadgerRepos, s.name)
+		from, err = badger.NewDatastore(path, &badger.DefaultOptions)
+		if err != nil {
+			return fmt.Errorf("connecting to badger source: %v", err)
+		}
+		log.Infof("connected to badger source: %s", path)
+	}
+	if len(toBadgerRepos) != 0 {
+		path := filepath.Join(toBadgerRepos, s.name)
+		if err := os.MkdirAll(path, os.ModePerm); err != nil {
+			return fmt.Errorf("making destination path: %v", err)
+		}
+		to, err = badger.NewDatastore(path, &badger.DefaultOptions)
+		if err != nil {
+			return fmt.Errorf("connecting to badger destination: %v", err)
+		}
+		log.Infof("connected to badger destination: %s", path)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	if len(fromMongoUri) != 0 {
+		uri, err := url.Parse(fromMongoUri)
+		if err != nil {
+			return fmt.Errorf("parsing source mongo URI: %v", err)
+		}
+		if len(fromMongoDatabase) == 0 {
+			return fmt.Errorf("source mongo database not specified")
+		}
+		from, err = mongods.New(ctx, fromMongoUri, fromMongoDatabase, mongods.WithCollName(s.name))
+		if err != nil {
+			return fmt.Errorf("connecting to mongo source: %v", err)
+		}
+		log.Infof("connected to mongo source: %s", uri.Redacted())
+	}
+	if len(toMongoUri) != 0 {
+		uri, err := url.Parse(toMongoUri)
+		if err != nil {
+			return fmt.Errorf("parsing destination mongo URI: %v", err)
+		}
+		if len(toMongoDatabase) == 0 {
+			return fmt.Errorf("destination mongo database not specified")
+		}
+		to, err = mongods.New(ctx, toMongoUri, toMongoDatabase, mongods.WithCollName(s.name))
+		if err != nil {
+			return fmt.Errorf("connecting to mongo destination: %v", err)
+		}
+		log.Infof("connected to mongo destination: %s", uri.Redacted())
+	}
+
+	res, err := from.Query(query.Query{})
+	if err != nil {
+		return fmt.Errorf("querying source: %v", err)
+	}
+	defer res.Close()
+
+	var lock sync.Mutex
+	var errors []string
+	var count int
+	start := time.Now()
+	lim := make(chan struct{}, parallel)
+	for r := range res.Next() {
+		if r.Error != nil {
+			return fmt.Errorf("getting next source result: %v", r.Error)
+		}
+		lim <- struct{}{}
+
+		r := r
+		go func() {
+			defer func() { <-lim }()
+
+			if err := to.Put(ds.NewKey(r.Key), r.Value); err != nil {
+				lock.Lock()
+				errors = append(errors, fmt.Sprintf("copying %s: %v", r.Key, err))
+				lock.Unlock()
+				return
+			}
+			if verbose {
+				log.Infof("copied %s", r.Key)
+			}
+			lock.Lock()
+			count++
+			lock.Unlock()
+			if count%parallel == 0 {
```

Comment:
Nit: move `lock.Unlock()` to include the if case since `count` is being read without lock and might be a data-race with other goroutine owning the lock. (tongue twister)

Reply:
Nice, done